### DISCUSSION TITLE
Need Help with Cross Compiling ImageMagick for arm64 on Android- Fully Static compiled libs

### DISCUSSION THREAD

**Discussion by ramasita** [MolotovCherry/Android-ImageMagick7#138] [Q&A] [answered]
Hi everyone,

I'm working on a project where I need to cross compile and build static libraries for Magick++, MagickCore, and MagickWand for arm64, to integrate with my existing Android project.

I've set up my environment variables and configured ImageMagick 7.1.1-25 for my host (aarch64-unknown-linux-android) and build (x86_64-pc-linux-gnu) systems. Here's a summary of my configuration:

Shared libraries: disabled
Static libraries: enabled
Various other options like OpenMP, security policy, and quantum depth are set as per requirements.
Libraries like JPEG, PNG, TIFF, etc., are enabled.
The issue arises when I run 'make'. I encounter build errors related to the __GNUC_PREREQ macro not being defined. It seems the problem is due to the inclusion of x86_64 specific headers in the CFLAGS. For instance, -I/usr/include/x86_64-linux-gnu is present, which shouldn't be for arm64.

I believe this incorrect path inclusion is causing the errors, but I'm not sure how to remove or correct these auto-included x86_64 paths in the CFLAGS.

Has anyone here successfully compiled fully static for ImageMagick targeting arm64 on Android if so could you please share it? 

I'd  also greatly appreciate any guidance or suggestions especially on how to fix the CFLAGS to be appropriate for arm64.

Thank you in advance for your help!

Summary of configure env for reference:
ImageMagick 7.1.1-25 is configured as follows. Verify that this configuration
matches your expectations.

Host system type: aarch64-unknown-linux-android
Build system type: x86_64-pc-linux-gnu

                Option                              Value
Shared libraries --enable-shared=no no
Static libraries --enable-static=yes yes
Build utilities --with-utilities=yes yes
Module support --with-modules=no no
GNU ld --with-gnu-ld=yes yes
OpenMP --enable-openmp= yes
Security policy --with-security-policy=open open
Quantum depth --with-quantum-depth=16 16
Channel mask depth --enable-64bit-channel-masks=no 32
High Dynamic Range Imagery
--enable-hdri=yes yes

Install documentation: yes

Memory allocation library:
JEMalloc --with-jemalloc=no no
MTMalloc --with-mtmalloc=no no
TCMalloc --with-tcmalloc=no no
UMem --with-umem=no no

Delegate library configuration:
BZLIB --with-bzlib=yes no
Autotrace --with-autotrace=no no
DJVU --with-djvu=yes yes
DPS --with-dps=no no
FFTW --with-fftw=no no
FLIF --with-flif=no no
FlashPIX --with-fpx=no no
FontConfig --with-fontconfig=yes yes
FreeType --with-freetype=yes yes
Ghostscript lib --with-gslib=no no
Graphviz --with-gvc=yes no
HEIC --with-heic=yes no
JBIG --with-jbig=yes no
JPEG v1 --with-jpeg=yes yes
JPEG XL --with-jxl=yes no
DMR --with-dmr=yes no
LCMS --with-lcms=yes yes
LQR --with-lqr=yes yes
LTDL --with-ltdl=no no
LZMA --with-lzma=yes yes
Magick++ --with-magick-plus-plus=yes yes
OpenEXR --with-openexr=yes yes
OpenJP2 --with-openjp2=yes yes
PANGO --with-pango=yes yes
PERL --with-perl=no no
PNG --with-png=yes yes
RAQM --with-raqm=yes no
RAW --with-raw=yes no
RSVG --with-rsvg=no no
TIFF --with-tiff=yes yes
WEBP --with-webp=yes no
WMF --with-wmf=no no
X11 --with-x= no
XML --with-xml=yes yes
ZIP --with-zip=yes no
ZLIB --with-zlib=yes yes
ZSTD --with-zstd=yes no

Delegate program configuration:
GhostPCL None pcl6 (unknown)
GhostXPS None gxps (unknown)
Ghostscript None gs (9.55.0)

Font configuration:
Apple fonts --with-apple-font-dir=default
Dejavu fonts --with-dejavu-font-dir=default none
Ghostscript fonts --with-gs-font-dir=default /usr/share/fonts/type1/gsfonts/
URW-base35 fonts --with-urw-base35-font-dir=default none
Windows fonts --with-windows-font-dir=default none

X11 configuration:
X_CFLAGS =
X_PRE_LIBS =
X_LIBS =
X_EXTRA_LIBS =

Options used to compile and link:
PREFIX = /usr/local
EXEC-PREFIX = /usr/local
VERSION = 7.1.1-25
CC = ...ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android24-clang
CFLAGS = -I/usr/include/libxml2 -I/usr/include/x86_64-linux-gnu -I/usr/include/libpng16 -pthread -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/OpenEXR -I/usr/include/lqr-1 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/openjpeg-2.4 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -pthread -fopenmp -Wall -g -O2 -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -DMAGICKCORE_CHANNEL_MASK_DEPTH=32
CPPFLAGS = -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -DMAGICKCORE_CHANNEL_MASK_DEPTH=32
PCFLAGS =
DEFS = -DHAVE_CONFIG_H
LDFLAGS =
LIBS = -llcms2 -ltiff -lfreetype -ljpeg -llqr-1 -lglib-2.0 -lpng16 -lz -ldjvulibre -lfontconfig -lfreetype -llzma -lIlmImf-2_5 -lImath-2_5 -lHalf-2_5 -lIex-2_5 -lIexMath-2_5 -lIlmThread-2_5 -lopenjp2 -lpangocairo-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lharfbuzz -lcairo -lxml2 -lz -lm -lomp
CXX = .../ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android24-clang++
CXXFLAGS = -g -O2 -pthread
FEATURES = Channel-masks(32-bit) Cipher DPC HDRI OpenMP
DELEGATES = djvu fontconfig freetype jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps tiff xml zlib

Error seen when I run 'make':
/usr/include/x86_64-linux-gnu/sys/cdefs.h:202:5: error: function-like macro '__GNUC_PREREQ' is not defined
#if __GNUC_PREREQ (4,3)

**Comment by MolotovCherry** [comment]
The build has not been tested on Linux (as per the [FAQ entry](https://github.com/MolotovCherry/Android-ImageMagick7/wiki/FAQ#it-wont-build-on-linux)), so I'd try using Windows to build it first of all. Alternatively, you could get github to build it for you, or try the docker image.

**Comment by ramasita** [comment]
Thank you for your response and guidance on my previous query. Following your advice, I initiated a fully static build from the ImageMagick GitHub repository. I used the --enable-static=yes option in the configure-any.sh script to ensure a static build.

However, I have a question regarding the static libraries. In the binaries attached to the repository, I noticed the absence of static libraries like libmagick++, libMagicCore, or libMagicWand. I'm trying to understand whether these static libraries are indeed part of the provided binaries, and perhaps I'm just not locating them correctly.

Could you please shed some light on this? Do these static libraries exist in the repo, or are they not included in the binaries for some reason? Any clarification on this would be very helpful.

**Reply by MolotovCherry** [reply]
The static libraries are built into the resulting binary in case of a static build (minus the 2 shared libs provided with it). In case of a shared lib build, the binary is dependent on the shared libraries of course

I suppose there was no real need to distribute the static libs, since the focus was making things easy to use, but they could be built from scratch if you really want them for some reason

**Reply by ramasita** [reply]
Thanks for the clarification on the static libraries. It's interesting to know that in a static build, the static libraries are incorporated directly into the resulting binary, with the exception of the two shared libraries provided alongside it.

With this understanding, I have a couple of follow-up questions:

1. Linking the 'magick' Binary in CMake: Given that the static libraries are embedded in the 'magick' binary, is it possible for me to link just this binary in my CMake project? Would this approach be sufficient for accessing all the functionalities of imagemagick that I need?

> I just checked the magick binary which is part of static binaries. Based on this below, the binary appears to be a **dynamically linked** ELF **executable** for ARM aarch64. I was under the impression that this was fully statically linked(.a) and it errors out on my end when I used this in my CMake.
```
/static_arm64-v8a] file magick 
magick: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), **dynamically linked**, interpreter /system/bin/linker64, BuildID[sha1]=db8719af844c99c0a579971e60d581ef864e803d, stripped
```

2. Building Static Libraries: It appears that my build process resulted in the creation of shared libraries instead of static ones - https://github.com/ramasita/Android-ImageMagick7/actions/runs/7340973657/artifacts/1136831517 .

Could you please share the specific options or steps to ensure a static library build? I'd like to understand how to configure the build process correctly to generate static libraries.

Thank you once again for your time and assistance.

**Reply by MolotovCherry** [reply]
1. Link the binary? The binary is meant for running (without need for other lib files, save those 2 included with the static build of course). You will need either the static libs or shared libs to link against.
2. The static libs are probably under the `build` folder. The build scripts don't check there since they just use the resulting output files. You can edit the [build.yml](https://github.com/MolotovCherry/Android-ImageMagick7/blob/e59258b7f118ff39c5da56a88c1635ae62a23f10/.github/workflows/build.yml#L43) file to upload another folder

> Based on this below, the binary appears to be a dynamically linked ELF executable for ARM aarch64.

It's dynamically linked against 2 shared libs (included in the folder)

**Reply by ramasita** [reply]
Thanks @MolotovCherry for your prompt response. 

I was able to build it successfully except I couldn't build the OpenMP as a static lib so I just disabled it.

**Reply by ramasita** [reply]
@MolotovCherry I'm seeing a linker issue when I build the debug version of my app's executable. These are all pointing to absence of zlib. Looks like ImageMagick depends on zlib for compression support. 

I checked the static libs that were built and confirmed there is no 'libz.a' built. Could you suggest me on how to resolve this error?

```
ld.lld: error: undefined symbol: gzerror
>>> referenced by blob.c:649 (././ImageMagick-7.1.1-24/MagickCore/blob.c:649)
>>>               blob.o:(CloseBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced by blob.c:3910 (././ImageMagick-7.1.1-24/MagickCore/blob.c:3910)
>>>               blob.o:(ReadBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced by blob.c:4928 (././ImageMagick-7.1.1-24/MagickCore/blob.c:4928)
>>>               blob.o:(ReadBlobString) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced 1 more times

ld.lld: error: undefined symbol: gzclose
>>> referenced by blob.c:725 (././ImageMagick-7.1.1-24/MagickCore/blob.c:725)
>>>               blob.o:(CloseBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a

ld.lld: error: undefined symbol: gzflush
>>> referenced by blob.c:5537 (././ImageMagick-7.1.1-24/MagickCore/blob.c:5537)
>>>               blob.o:(SyncBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a

ld.lld: error: undefined symbol: gzeof
>>> referenced by blob.c:1263 (././ImageMagick-7.1.1-24/MagickCore/blob.c:1263)
>>>               blob.o:(EOFBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced by blob.c:3914 (././ImageMagick-7.1.1-24/MagickCore/blob.c:3914)
>>>               blob.o:(ReadBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a

ld.lld: error: undefined symbol: gzseek
>>> referenced by blob.c:2927 (././ImageMagick-7.1.1-24/MagickCore/blob.c:2927)
>>>               blob.o:(IsBlobSeekable) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced by blob.c:5068 (././ImageMagick-7.1.1-24/MagickCore/blob.c:5068)
>>>               blob.o:(SeekBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a

ld.lld: error: undefined symbol: gzopen
>>> referenced by blob.c:3225 (././ImageMagick-7.1.1-24/MagickCore/blob.c:3225)
>>>               blob.o:(OpenBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced by blob.c:3225 (././ImageMagick-7.1.1-24/MagickCore/blob.c:3225)
>>>               blob.o:(OpenBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a

ld.lld: error: undefined symbol: gzread
>>> referenced by blob.c:3859 (././ImageMagick-7.1.1-24/MagickCore/blob.c:3859)
>>>               blob.o:(ReadBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a

ld.lld: error: undefined symbol: gzgetc
>>> referenced by blob.c:3873 (././ImageMagick-7.1.1-24/MagickCore/blob.c:3873)
>>>               blob.o:(ReadBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced by blob.c:3882 (././ImageMagick-7.1.1-24/MagickCore/blob.c:3882)
>>>               blob.o:(ReadBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced by blob.c:3891 (././ImageMagick-7.1.1-24/MagickCore/blob.c:3891)
>>>               blob.o:(ReadBlob) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
>>> referenced 1 more times

ld.lld: error: undefined symbol: gzgets
>>> referenced by blob.c:4924 (././ImageMagick-7.1.1-24/MagickCore/blob.c:4924)
>>>               blob.o:(ReadBlobString) in archive /Android-ImageMagick7/ImageMagick-7/lib/arm64-v8a/libmagickcore-7.a
```

Thanks!

**Reply by MolotovCherry** [reply]
Do you have the SDK/NDK set up properly (and you're building on Windows)? That file is located here:
`<NDKROOT>\toolchains\llvm\prebuilt\windows-x86_64\sysroot\usr\lib\aarch64-linux-android\libz.a`

The SDK/NDK requires things like env var setups.

Edit: Did you also follow the building instructions in the wiki?
https://github.com/MolotovCherry/Android-ImageMagick7/wiki/Setup--&--building-instructions#building-from-source

After that's done, you run the build-release script
https://github.com/MolotovCherry/Android-ImageMagick7/blob/master/build-release.bat

**Reply by ramasita** [reply]
I've successfully built the static libs using GitHub workflow. I was able to locate the libz.a from <NDKROOT>\toolchains\llvm\prebuilt\linux-x86_64\sysroot\usr\lib\aarch64-linux-android. I will try compiling along with this lib and confirm.

Thanks again for your prompt response.

**Comment by MrR0b0X** [comment, marked as answer]
@ramasita saw in your commits you were trying to build armeabi-v7a
i want to inform that the source now supports all cores (`arm64-v8a` | `armeabi-v7a` | `x86` | `x86_64`)

you can also trigger all for building all at once or the wanted core in local build/Handle release workflow's
![Screenshot 2025-02-12 035342](https://github.com/user-attachments/assets/5f0b29ab-6369-4f84-bf83-6aa3560ef2d6)
